torch/csrc/utils/future.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+
+namespace utils {
+
+// FutuerError inherits from std::exception, it can return const char* or
+// std::string error message
+struct TORCH_API FutureError final : public std::exception {
+  FutureError(std::string&& error_msg_) : error_msg(std::move(error_msg_)) {}
+
+  FutureError() = default;
+
+  const char* what() const noexcept override {
+    return error_msg.c_str();
+  }
+
+  std::string errMsg() const {
+    return error_msg;
+  }
+
+  std::string error_msg;
+};
+
+// This class holds a value of type T that will be ready in the future.
+// Most implementation is copied from FutureMessage and
+// c10::ivalue::Future
+template <typename T>
+class TORCH_API Future final {
+ public:
+  using Callback = std::function<void(const T&, bool, const FutureError&)>;
```

Comment:
Wondering aloud (genuinely): do we really want exception handling part of the future framework here, rather than part of the templated T class?

In the context of Future<Message>, adding the exception handling doesn't seem to help much, since exceptions are effectively part of Message class, as-is.

And exceptions aren't strictly needed in a future impl - do we really need more than std::future<> with addCallback support here?

Reply:
@jjlilley that is good point. I'm motivated to add error handling here mainly because c10::ivalue::future and folly::future have this. And we need to make it compatible with c10::ivalue::future, as eventually c10::ivalue::future will implement it as Future<IValue>. 

On the other hand, I feel like it will be convenient for general cases to have this error handling, although our FutureMessage do not benefit much from it (Message could be an exception type and we have a lot of call back error handling for rref and dist autograd).  Most of classes e.g., IValue types, they do not have error types and etc, it will be hard to handle these exceptions in those cases. 

overall I prefer to have error handing in future. How do you think? @mrshenli @pritamdamania  @pietern?

Reply:
Sorry, if there is an explicit goal to unify with ivalue::Future, I do see the parallelism here and the rationale for integrating the error handling. I don't have a strong objection here, and didn't mean to bring up gratuitously, I just wanted to clarify the requirements/constraints here, given that this widens the api.

(My impression was that previously there was some consensus that ivalue::Future and FutureMessage were separate enough - e.g. intrusive_ptr aspect, and easy enough to chain when we needed them to work together, that not a lot of upside in combining. But perhaps some day?)

Related question: if we go this direction, wondering if it makes more sense to use the function<void(void)> Callback being returned by ivalue::Future, rather than adding more params, since the lambda user might well just capture the future for that info?

Reply:
yes, we discussed with JIT team again, and the conclusion is to better have one Future in the long term, so eventually ivalue::Future should be merged. 

Ok, I will still keep error handling.